kernel/thread.c

```diff
@@ -751,6 +751,8 @@ FUNC_NORETURN void k_thread_user_mode_enter(k_thread_entry_t entry,
 	_current->entry.parameter3 = p3;
 #endif
 #ifdef CONFIG_USERSPACE
+	memset(_current->userspace_local_data, 0,
+	       sizeof(struct _thread_userspace_local_data));
```

Comment:
Question ? Couldn't it be optimized out by the compiler ? Shouldn't we add a barrier here, something like
` __asm__ __volatile__("" ::: "memory");`

Reply:
I'm not following you here. Why would this be optimized out?

This is the disassembly on x86:
```
Dump of assembler code for function k_thread_user_mode_enter:
   0x0010acb7 <+0>:	mov    0x130fc8,%eax
   0x0010acbc <+5>:	push   %ebp
   0x0010acbd <+6>:	mov    %esp,%ebp
   0x0010acbf <+8>:	orb    $0x4,0xc(%eax)
   0x0010acc3 <+12>:	call   0x10a4c5 <z_thread_essential_clear>
   0x0010acc8 <+17>:	push   $0x4
   0x0010acca <+19>:	mov    0x130fc8,%eax
   0x0010accf <+24>:	push   $0x0
   0x0010acd1 <+26>:	pushl  0x64(%eax)
   0x0010acd4 <+29>:	call   0x105090 <memset>
   0x0010acd9 <+34>:	add    $0xc,%esp
   0x0010acdc <+37>:	pushl  0x14(%ebp)
   0x0010acdf <+40>:	pushl  0x10(%ebp)
   0x0010ace2 <+43>:	pushl  0xc(%ebp)
   0x0010ace5 <+46>:	pushl  0x8(%ebp)
   0x0010ace8 <+49>:	call   0x1049bd <arch_user_mode_enter>
```

Reply:
There are compilers that, depending on optimizations flags, can remove memset() if the variable is not longer used. So in general, it's recommended to add a barrier when clearing out sensitive data. But I may be wrong.

https://wiki.sei.cmu.edu/confluence/display/c/MSC06-C.+Beware+of+compiler+optimizations
https://www.cryptologie.net/article/419/zeroing-memory-compiler-optimizations-and-memset_s/

Reply:
This buffer isn't on the call stack of the `k_thread_user_mode_enter()` function though...

Reply:
yep, examples there are just about a variables on the stack, so much likely this is not necessary and any compiler would remove it :)